p2p/src/net/default_backend/backend.rs

```diff
@@ -166,26 +169,27 @@ where
                         address,
                         error: P2pError::DialError(DialError::ConnectionRefusedOrTimedOut),
                     })
-                    .await
                     .map_err(P2pError::from)
             }
         }
     }
 
-    /// Disconnect remote peer by id
-    async fn disconnect_peer(&mut self, peer_id: &PeerId) -> crate::Result<()> {
+    /// Disconnect remote peer by id. Might fail if the peer is already disconnected.
+    fn disconnect_peer(&mut self, peer_id: &PeerId) -> crate::Result<()> {
         self.request_mgr.unregister_peer(peer_id);
-        if let Some(context) = self.peers.remove(peer_id) {
-            context.tx.send(Event::Disconnect).await.map_err(P2pError::from)
-        } else {
-            // TODO: Think about error handling. Currently we simply follow the libp2p behaviour.
-            log::error!("{peer_id} peer doesn't exist");
-            Ok(())
-        }
+
+        let peer = self
+            .peers
+            .remove(peer_id)
+            .ok_or(P2pError::PeerError(PeerError::PeerDoesntExist))?;
+
+        peer.tx.send(Event::Disconnect).map_err(P2pError::from)?;
+
+        Ok(())
```

Comment:
Could it be just this?

```suggestion
        peer.tx.send(Event::Disconnect).map_err(P2pError::from)
```

There are quite a few occurrences of this pattern all over the place but not sure my proposal is much better. Not a big difference.

Reply:
Sure, updated, thanks.